velox/expression/Expr.cpp

```diff
@@ -1149,75 +1149,83 @@ void Expr::evalWithMemo(
     VectorPtr& result) {
   VectorPtr base;
   distinctFields_[0]->evalSpecialForm(rows, context, base);
-  ++numCachableInput_;
-  if (baseDictionary_ == base) {
-    ++numCacheableRepeats_;
-    if (cachedDictionaryIndices_) {
-      LocalSelectivityVector cachedHolder(context, rows);
-      auto cached = cachedHolder.get();
-      VELOX_DCHECK(cached != nullptr);
-      cached->intersect(*cachedDictionaryIndices_);
-      if (cached->hasSelections()) {
-        context.ensureWritable(rows, type(), result);
-        result->copy(dictionaryCache_.get(), *cached, nullptr);
-      }
-    }
-    LocalSelectivityVector uncachedHolder(context, rows);
-    auto uncached = uncachedHolder.get();
-    VELOX_DCHECK(uncached != nullptr);
-    if (cachedDictionaryIndices_) {
-      uncached->deselect(*cachedDictionaryIndices_);
-    }
-    if (uncached->hasSelections()) {
-      // Fix finalSelection at "rows" if uncached rows is a strict subset to
-      // avoid losing values not in uncached rows that were copied earlier into
-      // "result" from the cached rows.
-      ScopedFinalSelectionSetter scopedFinalSelectionSetter(
-          context, &rows, uncached->countSelected() < rows.countSelected());
-
-      evalWithNulls(*uncached, context, result);
-      context.deselectErrors(*uncached);
-      context.exprSet()->addToMemo(this);
-      auto newCacheSize = uncached->end();
-
-      // dictionaryCache_ is valid only for cachedDictionaryIndices_. Hence, a
-      // safe call to BaseVector::ensureWritable must include all the rows not
-      // covered by cachedDictionaryIndices_. If BaseVector::ensureWritable is
-      // called only for a subset of rows not covered by
-      // cachedDictionaryIndices_, it will attempt to copy rows that are not
-      // valid leading to a crash.
-      LocalSelectivityVector allUncached(context, dictionaryCache_->size());
-      allUncached.get()->setAll();
-      allUncached.get()->deselect(*cachedDictionaryIndices_);
-      context.ensureWritable(*allUncached.get(), type(), dictionaryCache_);
-
-      if (cachedDictionaryIndices_->size() < newCacheSize) {
-        cachedDictionaryIndices_->resize(newCacheSize, false);
-      }
 
-      cachedDictionaryIndices_->select(*uncached);
+  if (base.get() != baseOfDictionaryRawPtr_ ||
```

Comment:
Just checking the equality of the raw pointer is not enough, because this is changed from a `shared_ptr` to `weak_ptr`, the vector can be reused and modified in place when the other place has `shared_ptr::unique` check pass.

Reply:
@Yuhta good catch. In that case, I can also hold onto the base as well on the second repeat instance, that way then I reuse the cached results on the third repeat I will atleast have the guarantee that the bases are now the same.

Reply:
right, to give you more context: the idea here is that in most cases the base is reused in multiple batches and caching it starting from the first instance provides little benefit as it will only potentially save one iteration but instead would end up holding onto the base and the cached results for every single input batch even in cases where its not required (like when the inputs dont have a repeating base).
With this change we would still retain the benefits of dict memoization while ensuring we dont hold only vector unnecessarily.